dali/kernels/signal/resampling_gpu.cuh

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+#define DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/signal/resampling.h"
+#include "dali/core/util.h"
+
+#define SHM_NCHANNELS 16
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace resampling {
+
+struct SampleDesc {
+  void *out;
+  const void *in;
+  ResamplingWindow window;
+  int64_t in_len;  // num samples in input
+  int64_t out_begin;  // output region-of-interest start
+  int64_t out_end;  // output region-of-interest end
+  int nchannels;  // number of channels
+  double scale;  // in_sampling_rate / out_sampling_rate
+};
+
+/**
+ * @brief Gets intermediate floating point representation depending on the input/output types
+ */
+template <typename Out, typename In>
+__device__ float ConvertInput(In in_val) {
+  if (std::is_unsigned<Out>::value && std::is_signed<In>::value) {
+    return (ConvertSatNorm<float>(in_val) + 1.0f) * 0.5f;
+  } else if (std::is_signed<Out>::value && std::is_unsigned<In>::value) {
+    return ConvertSatNorm<float>(in_val) * 2.0f - 1.0f;  // treat half-range as 0
+  } else {
+    return ConvertSatNorm<float>(in_val);  // just normalize
+  }
+}
+
+/**
+ * @brief Resamples 1D signal (single or multi-channel), optionally converting to a different data type.
+ *
+ * @param samples sample descriptors
+ */
+template <typename Out, typename In, bool SingleChannel = false>
```

Comment:
I'm not asking to split it now, but for my taste the `SingleChannel` optimization could have been done in a separate PR.